### PR TITLE
SAAS-4146: Added VPC Flow Logs Metric Alarm Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            if (!describeAlarms.data.length) {
+                helpers.addResult(results, 2,
+                    'No CloudWatch metric alarms found', region);
+                return rcb();
+            }
+
+            let alarms =  describeAlarms.data.find(alarm => alarm.MetricName === filters.metricTransformations[0].metricName);
+            if (alarms){
+                helpers.addResult(results, 0,
+                    'CloudWatch alarms are configured for the VPC Flow Logs', region);
```
Comment (sadeed12345) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:77]:
```suggestion
                    'CloudWatch alarms are configured for the VPC Flow Logs', 
                    region);
```
```suggestion
                    'CloudWatch alarms are configured for the VPC Flow Logs', region);
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                    'No CloudWatch metric alarms found', region);
+                return rcb();
+            }
+
+            let alarms =  describeAlarms.data.find(alarm => alarm.MetricName === filters.metricTransformations[0].metricName);
+            if (alarms){
+                helpers.addResult(results, 0,
+                    'CloudWatch alarms are configured for the VPC Flow Logs', 
+                    region);
+
```
Comment (sadeed12345) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:79]:
```suggestion
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            let alarms =  describeAlarms.data.find(alarm => alarm.MetricName === filters.metricTransformations[0].metricName);
+            if (alarms){
+                helpers.addResult(results, 0,
+                    'CloudWatch alarms are configured for the VPC Flow Logs', 
+                    region);
+            } else {
+                helpers.addResult(results, 2,
+                    'CloudWatch alarms are not configured for the VPC Flow Logs', region);
+
```
Comment (sadeed12345) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:82]:
```suggestion
                    'CloudWatch alarms are not configured for the VPC Flow Logs', 
                    region);
```

File: collectors/aws/collector.js [hunk 4]
```diff
@@ -160,7 +165,7 @@ var calls = {
             params: {
                 limit: 50 // The max available
             }
-        }
+        },
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:168]:
```suggestion
        }
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'VPC Flow Logs Metric Alarm',
+    category: 'CloudWatch',
+    domain: 'Compliance',
+    description: 'Ensures metric filters are setup for alarms by CloudWatch to detect any changes from VPC.',
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:8]:
```suggestion
    description: 'Ensures metric filters are setup for alarms by CloudWatch to detect any changes from VPC flow logs.',
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    more_info: 'A metric alarm watches a single CloudWatch metric or the result of a math expression based on CloudWatch metrics. The alarm performs one or more actions based on the value of the metric or expression relative to a threshold over a number of time periods. The action can be sending a notification to an Amazon SNS topic.',
+    recommended_action: '1. Create a log group in CloudWatch logs; 2. Create a metric filter for log group with Namespace=LogMetrics and metric Threshold=1; 3.Create CloudWatch alarm for the metric filter',
+    link: 'https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html',
+    apis: ['CloudWatchLogs:describeMetricFilters', 'CloudWatch:describeAlarms'],
+    settings: {
+        vpc_flow_log_group: {
+            name: 'CloudWatch VPC flow log group name',
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:15]:
```suggestion
            name: 'CloudWatch VPC Flow Log Group Name',
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                return rcb();
+            }
+
+            let filters = describeMetricFilters.data.find(metrics => metrics.logGroupName === config.vpc_flow_log_group);
+            if (!filters) {
+                helpers.addResult(results, 2,
+                    'No desired VPC group found', region);
+                return rcb();
+            }
+
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:57]:
```suggestion
            let cwVpcLogGroup = describeMetricFilters.data.find(metrics => metrics.logGroupName === config.vpc_flow_log_group);
            if (!cwVpcLogGroup) {
                helpers.addResult(results, 2,
                    'Unable to locate the specified log group', region);
                return rcb();
            }

```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            if (!describeAlarms.data.length) {
+                helpers.addResult(results, 2,
+                    'No CloudWatch metric alarms found', region);
+                return rcb();
+            }
+
+            let alarms =  describeAlarms.data.find(alarm => alarm.MetricName === filters.metricTransformations[0].metricName);
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:74]:
```suggestion
            let alarms =  describeAlarms.data.find(alarm => cwVpcLogGroup.metricTransformations && cwVpcLogGroup.metricTransformations.length && cwVpcLogGroup.metricTransformations[0].metricName && alarm.MetricName === cwVpcLogGroup.metricTransformations[0].metricName);
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            if (alarms){
+                helpers.addResult(results, 0,
+                    'CloudWatch alarms are configured for the VPC Flow Logs',
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:77]:
```suggestion
                    'CloudWatch alarm is configured for the VPC Flow Logs', 
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            }
+
+            let alarms =  describeAlarms.data.find(alarm => alarm.MetricName === filters.metricTransformations[0].metricName);
+            if (alarms){
+                helpers.addResult(results, 0,
+                    'CloudWatch alarms are configured for the VPC Flow Logs', 
+                    region);
+            } else {
+                helpers.addResult(results, 2,
+                    'CloudWatch alarms are not configured for the VPC Flow Logs',
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:81]:
```suggestion
                    'CloudWatch alarm is not configured for the VPC Flow Logs', 
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'VPC Flow Logs Metric Alarm',
+    category: 'CloudWatch',
+    domain: 'Compliance',
+    description: 'Ensures metric filters are setup for alarms by CloudWatch to detect any changes from VPC.',
+    more_info: 'A metric alarm watches a single CloudWatch metric or the result of a math expression based on CloudWatch metrics. The alarm performs one or more actions based on the value of the metric or expression relative to a threshold over a number of time periods. The action can be sending a notification to an Amazon SNS topic.',
+    recommended_action: '1. Create a log group in CloudWatch logs; 2. Create a metric filter for log group with Namespace=LogMetrics and metric Threshold=1; 3.Create CloudWatch alarm for the metric filter',
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:10]: Check these steps carefully

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    link: 'https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html',
+    apis: ['CloudWatchLogs:describeMetricFilters', 'CloudWatch:describeAlarms'],
+    settings: {
+        vpc_flow_log_group: {
+            name: 'CloudWatch VPC Flow Log Group Name',
+            description: 'Log group name for VPC to detect the log group being used by CLoudWatch logs',
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:16]:
```suggestion
            description: 'Existing CloudWatch log group name created to log VPC flow logs',
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.spec.js [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+};
+
+
+describe('vpcFlowLogsMetricAlarm', function () {
+    describe('run', function () {
+        it('should PASS if CloudWatch alarms are configured for the VPC Flow Logs', function (done) {
+            const cache = createCache([describeMetricFilters[0]], [describeAlarms[0]]);
+            vpcFlowLogsMetricAlarm.run(cache, { vpc_flow_log_group: 'vpc_flow_log_group_name'}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
```
Comment (AkhtarAmir) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.spec.js:146]: Add messages check in tests

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'VPC Flow Logs Metric Alarm',
+    category: 'CloudWatch',
+    domain: 'Compliance',
+    description: 'Ensure that CloudWatch group and alarm exist to detect any changes within VPC flow logs.',
+    more_info: 'A metric alarm watches a single CloudWatch metric or the result of a math expression based on CloudWatch metrics. The alarm performs one or more actions based on the value of the metric or expression relative to a threshold over a number of time periods. The action can be sending a notification to an Amazon SNS topic.',
+    recommended_action: '1. Create a log group in CloudWatch logs; 2. Create a metric filter for log group with Namespace=LogMetrics, metric Threshold=1 and check for logGroupName of desired type; 3.Create CloudWatch alarm for the metric filter',
```
Comment (sadeed12345) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:10]:
```suggestion
    recommended_action: '1. Create a log group in CloudWatch logs; 2. Create a metric filter for log group with Namespace=LogMetrics, metric Threshold=1 and check for logGroupName; 3.Create CloudWatch alarm for the metric filter',
```

File: plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'VPC Flow Logs Metric Alarm',
+    category: 'CloudWatch',
+    domain: 'Compliance',
+    severity: 'Low',
```
Comment (sadeed12345) [inline plugins/aws/cloudwatch/vpcFlowLogsMetricAlarm.js:8]:
```suggestion
    severity: 'LOW',
```

**Pull request merged**